_posts/2019-04-24-accessing-kafka-part-2.md

```diff
@@ -0,0 +1,317 @@
+
+But the node port services are just the infrastructure which can route the traffic to the brokers.
+We still need to configure the Kafka brokers to advertise the right address, so that the clients use this infrastructure.
+With node ports, the client connecting to the broker needs to connect to the:
+
+* address of one of the Kubernetes nodes
+* node port assigned to the service
+
+So Strimzi needs to gather these and configure these as the advertised addresses in the broker configuration.
+Strimzi is using separate listeners for the external and internal access.
+So any applications running inside the Kubernetes or OpenShift cluster will still use the old services and DNS names as described in part 1.
+
+Although node port services can route the traffic to the broker from all Kubernetes nodes, we can use only single address which will be advertised to the clients.
+And using the address of the actual node where the broker is running will mean less forwarding.
+But every time the broker restarts, the node might change.
+Therefore Strimzi is using an [init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) which is run every time when the Kafka broker pod is starting.
+It collects the address of the node and uses it to configure the advertised address.
+
+![Using init-container to get the node address]({{ "/assets/2019-04-24-init-container.png" }})
+
+To get the node address, the init container has to talk with the Kubernetes API and get the node resource.
+In the status of the node resource, the address is normally listed as one of the following:
+
+* External DNS
+* External IP
+* Internal DNS
+* Internal IP
+* Hostname
+
+Sometimes, only some of these are listed in the status.
+The init container will try to get one of them in the same order as they are listed above and will use the first one it finds.
+
+Once the address is configured, the client can use the bootstrap node port service to make the initial connection.
+From there the client will get the metadata containing the addresses of the individual brokers and start sending and receiving messages.
+
+![Client connecting to a broker using node port service]({{ "/assets/2019-04-24-client-connecting.png" }})
+
+After Strimzi configures everything inside the Kubernetes and Kafka clusters for you, you need to do just two things:
+
+* Get the node port number of the external bootstrap service (replace `my-cluster` with the name of your cluster):
+
+```
+kubectl get service my-cluster-kafka-external-bootstrap -o=jsonpath='{.spec.ports[0].nodePort}{"\n"}'
+```
+
+* Get the address of one of the nodes in your Kubernetes cluster (replace `node-name` with the name of one of your nodes - use `kubectl get nodes` to list all nodes):
+
+```
+kubectl get node node-name -o=jsonpath='{range .status.addresses[*]}{.type}{"\t"}{.address}{"\n"}'
+```
+
+The node address and node port number give you all what is needed to connect to your cluster:
```

Comment:
```suggestion
The node address and node port number give you all that is needed to connect to your cluster:
```

Reply:
Maybe you could just mention that you are going to use the `kafka-console-producer` tool to show that.